airwallex/src/main/java/com/airwallex/android/view/PaymentMethodsViewModel.kt

```diff
@@ -96,67 +86,69 @@ internal class PaymentMethodsViewModel(
-                    resultData.value = Result.failure(exception)
-                }
+        } catch (exception: AirwallexException) {
+            resultData.value = Result.failure(exception)
+            return
+        }
+        availablePaymentMethodPageNum.incrementAndGet()
+        availablePaymentMethodList.addAll(response.items ?: emptyList())
+        if (response.hasMore) {
+            retrieveAvailablePaymentMethods(
+                availablePaymentMethodList,
+                availablePaymentMethodPageNum,
+                resultData,
+                clientSecret
+            )
+        } else {
+            resultData.value =
+                getFilteredPaymentMethodsResult(availablePaymentMethodList)
+        }
+    }
 
-                override fun onSuccess(response: AvailablePaymentMethodTypeResponse) {
-                    availablePaymentMethodPageNum.incrementAndGet()
-                    availablePaymentMethodList.addAll(response.items ?: emptyList())
-                    if (response.hasMore) {
-                        retrieveAvailablePaymentMethods(
-                            availablePaymentMethodList,
-                            availablePaymentMethodPageNum,
-                            resultData,
-                            clientSecret
-                        )
-                    } else {
-                        when (session) {
-                            is AirwallexRecurringSession, is AirwallexRecurringWithIntentSession -> {
-                                resultData.value =
-                                    Result.success(
-                                        availablePaymentMethodList.filter {
-                                            it.transactionMode == TransactionMode.RECURRING
-                                        }.filter {
-                                            it.name !in unsupportedPaymentMethodTypes
-                                        }
-                                    )
-                            }
-                            is AirwallexPaymentSession -> {
-                                resultData.value =
-                                    Result.success(
-                                        availablePaymentMethodList.filter {
-                                            it.transactionMode == TransactionMode.ONE_OFF
-                                        }.filter {
-                                            it.name !in unsupportedPaymentMethodTypes
-                                        }
-                                    )
-                            }
-                            else -> {
-                                resultData.value =
-                                    Result.failure(AirwallexCheckoutException(message = "Not support session $session"))
-                            }
-                        }
+    private fun getFilteredPaymentMethodsResult(
```

Comment:
As discussed, it would be good if we can move this to Airwallex.kt to centralise the filtering